### ISSUE TITLE
fastq/reader can panic on invalid FASTQ files with mixed line endings

### ISSUE THREAD

**Issue by juliangehring** [zaeleus/noodles#166]
Certain cases of invalid FASTQ files with mixed line endings can trigger panics in `fastq::Reader` instead of throwing an error.

Two minimal examples (with line endings shown), run through the `fastq_count` example:

> @{␊
A␊
+␊
␊
@␊
+␍

> @␊
A␊
+␊
+␍

For example, the latter example triggers the following panic:

> thread 'main' panicked at 'range end index 18446744073709551615 out of range for slice of length 7', noodles/noodles-fastq/src/reader/record/definition.rs:46:30

I suspect that [`i - 1`](https://github.com/zaeleus/noodles/blob/master/noodles-fastq/src/reader/record/definition.rs#L46 ) gets negative in these edge cases, causing the panic.

To be clear, I wouldn't expect these files to be parsed successfully. However, invalid FASTQ files or mixed line endings (due to people editing them manually) are unfortunately a reality and throwing errors in all cases would be helpful.

**Comment by zaeleus** [comment]
Thanks for report! There definitely shouldn't be any panics. This issue is fixed in 8ed321e35d4be12d9d82d9abde34c95e03fecb43.

I'll note that the line readers in noodles follow the same rules as [`std::io::Lines`](https://doc.rust-lang.org/std/io/struct.Lines.html), i.e., they only handles `LF` and `CRLF` as newlines. This means that, in your example, the single trailing carriage return (`CR`) is included as part of the last line.

```rust
use std::io::{self, BufRead};

fn main() -> io::Result<()> {
    let data = b"@\nA\n+\n+\r";
    let reader = &data[..];
    let lines: Vec<_> = reader.lines().collect::<Result<_, _>>()?;
    dbg!(lines); // => ["@", "A", "+", "+\r"]
    Ok(())
}
```

**Comment by juliangehring** [comment]
Thank you for the ultra fast fix! Looks excellent.

**Comment by zaeleus** [comment]
This fix is now included in noodles 0.38.0/noodles-fastq 0.7.1.
